restful.entity.inc

```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_id')
+    ->propertyCondition('value', $entity_id)
+    ->propertyCondition('hash', $hashes, 'IN');
+  return CacheFragmentController::lookUpHashes($query);
+}
+
+/**
  * Implements hook_entity_update().
  */
 function restful_entity_update($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
+      ->getCacheController()
+      ->clear($hash);
+  }
 }
 
 /**
  * Implements hook_entity_delete().
  */
 function restful_entity_delete($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
+      ->getCacheController()
+      ->clear($hash);
+  }
 }
 
 /**
  * Implements hook_user_update().
  */
 function restful_user_update(&$edit, $account, $category) {
-  // Due to the limitations for cid matching on clearing caches, we need to
-  // clear all the bin. We cannot do $cid = '%::uu' . $account->uid . '::pa';
-  CacheDecoratedDataProvider::invalidateEntityCache('*');
+  // Search for all the cache fragments with our entity id.
+  $query = new \EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'user_id')
+    ->propertyCondition('value', $account->uid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (CacheFragmentController::lookUpHashes($query) as $hash) {
+    if (!$plugin_id = CacheFragmentController::resourceIdFromHash($hash)) {
+      continue;
+    }
+    try {
+      $handler = $resource_manager->getPlugin($plugin_id);
+    }
+    catch (PluginNotFoundException $e) {
+      continue;
+    }
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    // You can get away without the fragments for a clear.
+    $cache_object = new RenderCache(new ArrayCollection(), $hash, $handler->getCacheController());
```

Comment:
Instead of searching for all the `restful_cache_fragment` entities that reference this cache and delete them + clearing cache, we create the `RenderCache` object (that contains the cache) and call the `->clear()` method on it.